bindings/rust/s2n-tls-hyper/src/stream.rs

```diff
@@ -0,0 +1,100 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use hyper::rt::{Read, ReadBufCursor, Write};
+use hyper_util::{
+    client::legacy::connect::{Connected, Connection},
+    rt::TokioIo,
+};
+use s2n_tls::connection::Builder;
+use s2n_tls_tokio::TlsStream;
+use std::{
+    io::Error,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+/// `MaybeHttpsStream` is a wrapper over a hyper TCP stream, T, allowing for TLS to be negotiated
+/// over the TCP stream.
+///
+/// While not currently implemented, the `MaybeHttpsStream` enum will provide an `Http` type
+/// corresponding to the plain TCP stream, allowing for HTTP to be negotiated in addition to HTTPS
+/// when the HTTP scheme is used.
+///
+/// This struct is used to implement `tower_service::Service` for `HttpsConnector`, and shouldn't
+/// need to be used directly.
+pub enum MaybeHttpsStream<T, B>
+where
+    T: Read + Write + Unpin,
+    B: Builder,
+    <B as Builder>::Output: Unpin,
+{
+    // T is the underlying hyper TCP stream, which is wrapped in a `TokioIo` type in order to make
+    // it compatible with tokio (implementing AsyncRead and AsyncWrite). This allows the TCP stream
+    // to be provided to the `s2n_tls_tokio::TlsStream`.
+    //
+    // `MaybeHttpsStream` MUST implement hyper's `Read` and `Write` traits. So, the `TlsStream` is
+    // wrapped in an additional `TokioIo` type, which already implements the conversion from hyper's
+    // traits to tokio's. This allows the `Read` and `Write` implementations for `MaybeHttpsStream`
+    // to simply call the `TokioIo` `poll` functions.
+    Https(TokioIo<TlsStream<TokioIo<T>, B::Output>>),
+}
+
+impl<T, B> Connection for MaybeHttpsStream<T, B>
```

Comment:
Could we use a `use as HyperConnection` import to make this a little bit less ambiguous? Just so that s2n-tls and hyper connections don't get mixed up. https://doc.rust-lang.org/rust-by-example/mod/use.html
```suggestion
impl<T, B> HyperConnection for MaybeHttpsStream<T, B>
```